src/Storages/IStorageDataLake.h

```diff
@@ -0,0 +1,113 @@
+#pragma once
+
+#include "config.h"
+
+#if USE_AWS_S3
+
+#    include <Common/logger_useful.h>
+#    include <Storages/IStorage.h>
+
+#    include <filesystem>
+
+#    include <fmt/format.h>
+
+#    include <IO/S3/URI.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int NOT_IMPLEMENTED;
+}
+
+template <typename Storage, typename Name, typename MetadataParser>
+class IStorageDataLake : public Storage
+{
+public:
+    using Configuration = typename Storage::Configuration;
+    // 1. Parses internal file structure of table
+    // 2. Finds out parts with latest version
+    // 3. Creates url for underlying StorageS3 enigne to handle reads
+    IStorageDataLake(
+        const typename Storage::Configuration & configuration_,
+        const StorageID & table_id_,
+        ColumnsDescription columns_,
+        const ConstraintsDescription & constraints_,
+        const String & comment,
+        ContextPtr context_,
+        std::optional<FormatSettings> format_settings_)
+        : Storage(
+            getAdjustedConfiguration(
+                context_, Storage::updateConfiguration(context_, configuration_), &Poco::Logger::get("Storage" + String(name))),
+            table_id_,
+            columns_,
+            constraints_,
+            comment,
+            context_,
+            format_settings_)
+    {
+    }
+
+    static constexpr auto name = Name::name;
+    String getName() const override { return name; }
+
+    static ColumnsDescription getTableStructureFromData(
+        typename Storage::Configuration & configuration, const std::optional<FormatSettings> & format_settings, ContextPtr ctx)
+    {
+        Storage::updateConfiguration(ctx, configuration);
+
+        auto new_configuration = getAdjustedConfiguration(ctx, configuration, &Poco::Logger::get("Storage" + String(name)));
+
+        return Storage::getTableStructureFromData(new_configuration, format_settings, ctx, /*object_infos*/ nullptr);
+    }
+
+    static typename Storage::Configuration
+    getAdjustedConfiguration(const ContextPtr & context, const typename Storage::Configuration & configuration, Poco::Logger * log)
+    {
+        MetadataParser parser{configuration, context};
+
+        auto keys = parser.getFiles();
+        String new_uri = std::filesystem::path(configuration.url.uri.toString()) / Name::data_directory_prefix
+            / MetadataParser::generateQueryFromKeys(keys, configuration.format);
+
+        typename Storage::Configuration new_configuration(configuration);
```

Comment:
you already have `using Configuration = typename Storage::Configuration` in this class
```suggestion
        Configuration new_configuration(configuration);
```
and in other places in this file